src/sync.h

```diff
@@ -324,4 +559,83 @@ struct CLockLocation
 
 typedef std::vector<std::pair<void *, CLockLocation> > LockStack;
 typedef std::map<std::pair<void *, void *>, LockStack> LockStackMap;
+
+/** A thread corral is a granular thread organization technique.
+Code is assigned to a corral via Enter(...) and Exit(...) APIs (but use the scoped CCorralLock object instead of direct
+calls).
+Multiple threads can be running in the same corral, but threads cannot run in multiple corrals simultaneously.
+
+Higher corral numbers block lower ones, but are themselves blocked from entry until the all other corrals are clear.
+For example, lets us assume threads are running in corral region 1.
+If a thread wants to enter corral region 2, threads are blocked from entering region 1.  Once all threads have left 1,
+the thread(s) waiting to enter 2 are released.
+Now, a thread wants to enter corral region 1.  Threads can continue to enter and leave region 2 (since it is > 1).
+If all threads leave region 2, the threads waiting for region 1 are released.
```

Comment:
The way I read it is that if there are many low priority tasks running, and a high priority task wants to run then it has to wait for the lower priority threads to finish... but shouldn't any higher priority tasks be immediately allowed to run while the lower prioritiy tasks are prevented from enter the lower priority corral, rather than having to wait?

Reply:
yes lower priority tasks are prevented from entering, but we still have to wait for the ones currently in there to leave.